File: backend/app/schemas/workflow_operations.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


class WorkflowOperationReasonRequest(BaseModel):
  reason: str = Field(min_length=3, max_length=500)


class WorkflowIncidentUpdateRequest(WorkflowOperationReasonRequest):
  status: Literal["resolved", "ignored"]


class WorkflowOperationActionRead(BaseModel):
  id: UUID
  status: str
  manual_replay_count: int | None = None


class WorkflowOperationsIssueRead(BaseModel):
  category: str
  severity: str
  instance_id: UUID | None
  node_instance_id: UUID | None
  title: str
  message: str
  age_seconds: int | None


class WorkflowOutboxOperationRead(BaseModel):
  id: UUID
  instance_id: UUID
  node_instance_id: UUID | None
  event_type: str
  status: str
  attempt_count: int
  available_at: datetime | None
  last_error: str | None
  manual_replay_count: int
  last_replayed_at: datetime | None
  last_replayed_by_user_id: UUID | None
  last_replay_reason: str | None
  updated_at: datetime


class WorkflowIncidentOperationRead(BaseModel):
  id: UUID
  category: str
  status: str
  severity: str
  occurrence_count: int
  first_seen_at: datetime
  last_seen_at: datetime
  resolved_at: datetime | None
  resolved_by_user_id: UUID | None
  resolution_note: str | None
  instance_id: UUID | None
  node_instance_id: UUID | None
  task_id: UUID | None
  command_receipt_id: UUID | None
  outbox_event_id: UUID | None
  engine_version: str | None
  detail_keys: list[str] = Field(default_factory=list)


class WorkflowProjectionStreamHealthRead(BaseModel):
  stream_name: str
  status: str
  processed_count: int
  source_count: int
  backlog_count: int
  lag_seconds: int
  last_success_at: datetime | None
  last_error: str | None
  sampled_at: datetime


class WorkflowShadowHealthRead(BaseModel):
  scan_id: UUID
  sample_mode: str
  observed_at: datetime
  outcomes: dict[str, int] = Field(default_factory=dict)
  severities: dict[str, int] = Field(default_factory=dict)


class WorkflowOperationsMetricsRead(BaseModel):
  runs: dict[str, int] = Field(default_factory=dict)
  stalled_run_count: int
  suspended_node_count: int
  join_wait_count: int
  oldest_join_wait_seconds: int
  outbox: dict[str, int] = Field(default_factory=dict)
  outbox_backlog_count: int
  oldest_outbox_backlog_seconds: int
  projection_failed_stream_count: int


class WorkflowOperationsDashboardRead(BaseModel):
  generated_at: datetime
  stalled_minutes: int
  metrics: WorkflowOperationsMetricsRead
  projection_streams: list[WorkflowProjectionStreamHealthRead] = Field(default_factory=list)
  shadow: WorkflowShadowHealthRead | None
  issues: list[WorkflowOperationsIssueRead] = Field(default_factory=list)
  failed_outbox: list[WorkflowOutboxOperationRead] = Field(default_factory=list)
  incidents: list[WorkflowIncidentOperationRead] = Field(default_factory=list)


class WorkflowTraceRead(BaseModel):
  id: UUID
  event_type: str
  occurred_at: datetime
  request_id: str | None
  command_id: str | None
  correlation_id: UUID | None
  instance_id: UUID
  node_instance_id: UUID | None
  task_id: UUID | None
  actor_user_id: UUID | None
  payload_keys: list[str] = Field(default_factory=list)


class WorkflowTraceListRead(BaseModel):
  items: list[WorkflowTraceRead] = Field(default_factory=list)

File: frontend/src/types/workflowOperations.ts
export type WorkflowOperationAction = {
  id: string
  status: string
  manual_replay_count?: number | null
}

export type WorkflowOperationsIssue = {
  category: string
  severity: string
  instance_id: string | null
  node_instance_id: string | null
  title: string
  message: string
  age_seconds: number | null
}

export type WorkflowOutboxOperation = {
  id: string
  instance_id: string
  node_instance_id: string | null
  event_type: string
  status: string
  attempt_count: number
  available_at: string | null
  last_error: string | null
  manual_replay_count: number
  last_replayed_at: string | null
  last_replayed_by_user_id: string | null
  last_replay_reason: string | null
  updated_at: string
}

export type WorkflowIncidentOperation = {
  id: string
  category: string
  status: string
  severity: string
  occurrence_count: number
  first_seen_at: string
  last_seen_at: string
  resolved_at: string | null
  resolved_by_user_id: string | null
  resolution_note: string | null
  instance_id: string | null
  node_instance_id: string | null
  task_id: string | null
  command_receipt_id: string | null
  outbox_event_id: string | null
  engine_version: string | null
  detail_keys: string[]
}

export type WorkflowProjectionStreamHealth = {
  stream_name: string
  status: string
  processed_count: number
  source_count: number
  backlog_count: number
  lag_seconds: number
  last_success_at: string | null
  last_error: string | null
  sampled_at: string
}

export type WorkflowShadowHealth = {
  scan_id: string
  sample_mode: string
  observed_at: string
  outcomes: Record<string, number>
  severities: Record<string, number>
}

export type WorkflowOperationsDashboard = {
  generated_at: string
  stalled_minutes: number
  metrics: {
    runs: Record<string, number>
    stalled_run_count: number
    suspended_node_count: number
    join_wait_count: number
    oldest_join_wait_seconds: number
    outbox: Record<string, number>
    outbox_backlog_count: number
    oldest_outbox_backlog_seconds: number
    projection_failed_stream_count: number
  }
  projection_streams: WorkflowProjectionStreamHealth[]
  shadow: WorkflowShadowHealth | null
  issues: WorkflowOperationsIssue[]
  failed_outbox: WorkflowOutboxOperation[]
  incidents: WorkflowIncidentOperation[]
}

export type WorkflowTrace = {
  id: string
  event_type: string
  occurred_at: string
  request_id: string | null
  command_id: string | null
  correlation_id: string | null
  instance_id: string
  node_instance_id: string | null
  task_id: string | null
  actor_user_id: string | null
  payload_keys: string[]
}

export type WorkflowTraceList = {
  items: WorkflowTrace[]
}

export type WorkflowTraceFilters = Partial<{
  request_id: string
  command_id: string
  correlation_id: string
  instance_id: string
  node_instance_id: string
  task_id: string
}>
